### PR TITLE
Added Rancher Metadata backend

### PR DESCRIPTION
**Pull request by cloudnautique**
This adds backend support for Rancher Labs Rancher Metadata service.
It is meant to be used in a Rancher environment.

### CODE REVIEW

File: src/github.com/kelseyhightower/confd/backends/rancher/client.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+	log "github.com/kelseyhightower/confd/log"
+)
+
+const (
+	MetaDataURL = "http://rancher-metadata/2015-07-25"
+)
+
+type Client struct {
+	Url        string
```
Comment (ibuildthecloud) [inline src/github.com/kelseyhightower/confd/backends/rancher/client.go]: Do these need to be public? `Url` and `HttpClient`

File: src/github.com/kelseyhightower/confd/backends/rancher/client.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+	HttpClient *http.Client
+}
+
+func NewRancherClient(backendNodes []string) (*Client, error) {
+	url := MetaDataURL
+
+	if len(backendNodes) > 0 {
+		url = strings.Join([]string{"http://", backendNodes[0], "/2015-07-25"}, "")
```
Comment (ibuildthecloud) [inline src/github.com/kelseyhightower/confd/backends/rancher/client.go]: This is more straight forward as

`"http://" + b + "/2015-7-25"`

or

`fmt.Sprintf("http://%s/2015-07-25", backendNodes[0])`

File: src/github.com/kelseyhightower/confd/backends/rancher/client.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+	}
+
+	log.Info(strings.Join([]string{"Using Rancher Metadata URL: ", url}, ""))
```
Comment (ibuildthecloud) [inline src/github.com/kelseyhightower/confd/backends/rancher/client.go]: Just do `log.Info("Using Rancher Metadata URL: ", url)` or `log.Infof("Using Rancher Metadata URL: %s", url)` depending on the logging API

File: src/github.com/kelseyhightower/confd/backends/rancher/client.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+	client := &Client{
+		Url:        url,
+		HttpClient: &http.Client{},
+	}
+
+	err := client.testConnection()
```
Comment (ibuildthecloud) [inline src/github.com/kelseyhightower/confd/backends/rancher/client.go:37]: shorten to

```
if err := client.testConnection(); err != nil {
  return err
}
```

or

```
err := client.testConnection()
return client, err
```

File: src/github.com/kelseyhightower/confd/backends/rancher/client.go [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+	if err != nil {
+		return client, err
+	}
+
+	return client, nil
+}
+
+func (c *Client) GetValues(keys []string) (map[string]string, error) {
+	vars := make(map[string]string)
```
Comment (ibuildthecloud) [inline src/github.com/kelseyhightower/confd/backends/rancher/client.go]: for whatever reason this syntax is more popular

`vars := map[string]string{}`

File: src/github.com/kelseyhightower/confd/backends/rancher/client.go (hunk 3 above)
Comment (cloudnautique) [inline src/github.com/kelseyhightower/confd/backends/rancher/client.go]: I'll have to look again, but its not straight logrus.

File: src/github.com/kelseyhightower/confd/backends/rancher/client.go [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+	for _, key := range keys {
+		body, err := c.makeMetaDataRequest(key)
+		if err != nil {
+			return vars, err
+		}
+
+		var jsonResponse interface{}
+		err = json.Unmarshal(body, &jsonResponse)
+		if err != nil {
```
Comment (ibuildthecloud) [inline src/github.com/kelseyhightower/confd/backends/rancher/client.go]: where possible always short to the format

```
if err := doStuff(); err != nil {
  return err
}
```

File: src/github.com/kelseyhightower/confd/backends/rancher/client.go [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+	req, _ := http.NewRequest("GET", strings.Join([]string{c.Url, path}, ""), nil)
+	req.Header.Set("Accept", "application/json")
+
+	resp, err := c.HttpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
```
Comment (ibuildthecloud) [inline src/github.com/kelseyhightower/confd/backends/rancher/client.go]: shorten to `return ioutil.ReadAll(resp.Body)`

File: src/github.com/kelseyhightower/confd/backends/rancher/client.go [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+	if err != nil {
+		return nil, err
+	}
+	return body, nil
+}
+
+func (c *Client) testConnection() error {
+	var err error
+	max_time := 5 * time.Second
```
Comment (ibuildthecloud) [inline src/github.com/kelseyhightower/confd/backends/rancher/client.go]: change to camelCase, so maxTime

File: src/github.com/kelseyhightower/confd/backends/rancher/client.go [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+	for i := 1 * time.Second; i < max_time; i *= time.Duration(2) {
+		_, httpError := c.makeMetaDataRequest("/")
```
Comment (ibuildthecloud) [inline src/github.com/kelseyhightower/confd/backends/rancher/client.go]: can you shorten to

```
if _, err = c.makeMetaDataRequest("/"); err != nil {
  time.Sleep(1)
} else {
  return nil
}
```

**Pull request closed without merging**